### PR TITLE
Migrate ASM test to golang testing framework

### PR DESCRIPTION
**Pull request by ScottRMalley**
This PR migrates the ASM tests from python to our golang testing framework.

It includes:
* Upstream commit to allow error decoding
* Updated abigen template to generate custom errors from abi spec
* All current tests in /autonity/solidity/test/asm migrated

The primary ToDo, outside the scope of this PR will be to include testing for emitted events, once we have implemented support in the Golang framework.

### CODE REVIEW

File: accounts/abi/bind/source.java.tpl
Comment (yazzaoui) [inline accounts/abi/bind/source.java.tpl:1]: Do we really need that?

File: accounts/abi/abi.go [hunk 1]
```diff
@@ -261,6 +278,27 @@ func (abi *ABI) HasReceive() bool {
...
 
+// panicSelector is a special function selector for panic reason unpacking.
+var panicSelector = crypto.Keccak256([]byte("Panic(uint256)"))[:4]
+
+// panicReasons map is for readable panic codes
+// see this linkage for the details
+// https://docs.soliditylang.org/en/v0.8.21/control-structures.html#panic-via-assert-and-error-via-require
+// the reason string list is copied from ether.js
+// https://github.com/ethers-io/ethers.js/blob/fa3a883ff7c88611ce766f58bdd4b8ac90814470/src.ts/abi/interface.ts#L207-L218
+var panicReasons = map[uint64]string{
```
Comment (yazzaoui) [inline accounts/abi/abi.go:289]: Good idea👍

**Review by yazzaoui — APPROVED** [review]
LGTM, the new error testing mechanism is neat

File: accounts/abi/abi.go (hunk 1 above)
Comment (ScottRMalley) [reply accounts/abi/abi.go:289]: Wish I could say this one was me, but I pulled it from upstream

**Review by vpiyush — APPROVED** [review]
Looks Good to me!

**Pull request merged**